### PR TITLE
WirePlugin Improvements

### PR DESCRIPTION
**Pull request by oldergod**
- New: creates a new generation task per source (main, variants, etc)
- New: Adds wire-runtime automaticallly when there are JVM targets
- New: Adds generated code directories into sourceSets automatically

Includes support for Java, Kotlin (MPP), and Android.
Highly inspired from SqlDelight applied to Wire with the addition that Wire needs to support Java only projects.

It misses some tests, specially around Android but I had trouble setting up an Android project in the plugin tests so I'd rather go on with that now and follow up.
I tried on Cash and everything looks good.

- Are generated directory source aware? Nope. We can fix that in followup.

### CODE REVIEW

File: gradle.properties
Comment (swankjesse) [inline gradle.properties:2]: oooooooh is this the fix I’ve been needing?!

File: wire-grpc-sample/server/build.gradle
Comment (swankjesse) [inline wire-grpc-sample/server/build.gradle:16]: Nice.

File: wire-library/dependencies.gradle
Comment (swankjesse) [inline wire-library/dependencies.gradle:28]: I don’t understand these two... comment?

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (swankjesse) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:44]: explode if this is already initialized?

Comment (swankjesse) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:44]: (I guess I should read what the rules are for this)

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (swankjesse) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:65]: is this list of plugin IDs listed somewhere?

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (swankjesse) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:86]: nice

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (swankjesse) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:102]: this wants to be `check()`

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (swankjesse) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:107]: NICE

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (swankjesse) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:111]: FD?

**Review by swankjesse — COMMENTED** [review]
Still working through this....

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (JakeWharton) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:44]: The line above this will already throw if it's called twice.

It's never called twice, though.

File: wire-library/dependencies.gradle
Comment (AlecStrong) [reply wire-library/dependencies.gradle:28]: you should be able to `compileOnly` on `plugins.android` and not need `androidDependency`, if thats not the case per jesses comment would be good to have a comment here describing why

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt
Comment (AlecStrong) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt:49]: this looks familiar

**Review by AlecStrong — APPROVED** [review]

File: gradle.properties
Comment (oldergod) [reply gradle.properties:2]: Don't think so?

File: wire-library/dependencies.gradle
Comment (oldergod) [reply wire-library/dependencies.gradle]: I remove the redundancy. Since we are compile only we do whatever we want without forcing deps on consumers (no Android deps on non-android project).
I also updated to `4.0.1`

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:44]: I'll leave it as is.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:65]: I don't think so. I use what SqlDelight is using + the 2 javas. I looked around but didn't find anything.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:102]: fixed

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:111]: reverted.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (jrodbx) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:34]: why AtomicBoolean?  I don't typically see this used in Gradle plugins, is this a sign of a race condition we should otherwise avoid?

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (jrodbx) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:120]: use incremental task outputs for this?

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (jrodbx) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:157]: i don't think we should using sourceSets in task names.

a Gradle build target maps to a Gradle variant (which I believe inspired by the Android variant system).  

The variant helps decide what source sets to use, as a sort of implementation detail.

Comment (jrodbx) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:157]: think assemble -> assembleDebug, assembleRelease, not assembleMain.

File: wire-library/build.gradle
Comment (swankjesse) [inline wire-library/build.gradle:34]: What’s in JCenter?

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (swankjesse) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:209]: Nice

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt
Comment (swankjesse) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt:43]: Should we create a copy-pastable gist like `CommonEnvironments.kt` that has all this shared knowledge and no Wire knowledge. Something we could keep in sync with SQLDelight, and that we could eventually librarify if we ever needed a 3rd and 4th Gradle plugin with this stuff.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt
Comment (swankjesse) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt:151]: yes please

Comment (swankjesse) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt:151]: (Follow up!)

**Review by swankjesse — APPROVED** [review]

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt
Comment (jrodbx) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt:153]: is this coupling compilation to the generateProtos task?

if so, i feel pretty strongly on having this be the default.  our build speeds are terrible and re-generating protos in their modules constantly as part of the compile phase (gw app:assembleDebug) is a huge part of this.

**Comment by swankjesse** [comment]
This is awesome, and it fixes a problem I’ve had in some protos use I’m managing. I’d like to land this ASAP so I can unblock that work.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:157]: There would not be assembleMain on Android, only Debug and Release.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:120]: Can do in followup

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:34]: I am not aware of why, I was inspired by SqlDelight here

File: wire-library/build.gradle
Comment (oldergod) [reply wire-library/build.gradle:34]:
```
> Could not resolve all files for configuration ':wire-gradle-plugin:compileClasspath'.
   > Could not find org.jetbrains.trove4j:trove4j:20160824.
     Searched in the following locations:
       - ...
     Required by:
         project :wire-gradle-plugin > com.android.tools.build:gradle:4.0.1 > com.android.tools.build:builder:4.0.1 > com.android.tools:sdk-common:27.0.1
```

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt:153]: Our task should not run twice in a row. If there are cases when we run while it doesn't feel necessary we should fix. I think that's nice to have it as a task dependency.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (Egorand) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:157]: Would a task in a plain Kotlin JVM project be called `generateMainProtos` or `generateProtos`? My preference is for the latter.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (Egorand) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:202]: Is this comment wrong? In Kotlin multiplatform projects, this code would add the runtime dependency to the `commonMain` source set, which is expected. The project may not have JVM targets though, which is also fine.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (Egorand) [inline wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:215]:
```suggestion
        } catch (_: UnknownConfigurationException) {
```

**Review by Egorand — APPROVED** [review]

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:85]: I don't think it is actually. I'll rephrase.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:202]: I'll rephrase

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:157]: `generateProtos` would always be there whatever setup there is. It will execute all sources tasks.
Sor a plain JVM project would have
`generateProtos` and `generateMainProtos`.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt:153]: I think the brave strategy is too bind our task to the compilation and then to fix anything that sucks in ours.

File: sample/build.gradle
Comment (oldergod) [reply sample/build.gradle:2]: removing, you're right! Some leftover of my iterations.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (oldergod) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:215]: fixed

**Comment by oldergod** [comment]
> This is awesome, and it fixes a problem I’ve had in some protos use I’m managing. I’d like to land this ASAP so I can unblock that work.

@jrodbx wants some time to think this over before merging. Would their be anything we wouldn't be able to iterate/fix/revert later if we don't like how things are?

**Comment by oldergod** [comment]
@jrodbx wanna block the variant part so I'll take it out of the PR so we can merge.

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt
Comment (jrodbx) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt:34]: I'm reluctant to copy the SqlDelight implementation here.

How about not merging in the variant code in this PR?

We'd still accomplish 2 of the 3 objectives:
New: Adds wire-runtime automaticallly when there are JVM targets
New: Adds generated code directories into sourceSets automatically

File: wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt
Comment (jrodbx) [reply wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt:153]: > Since generateProtos is basically a compilation task

No, it isn't, it's a code generation task.  We combine its output with other source files to proceed to compilation phase.  Perhaps I'm missing semantics.

> I think the brave strategy is too bind our task to the compilation and then to fix anything that sucks in ours.

We say that now, but our team is larger now and still growing (which is awesome!) so this affects more people.

**Comment by oldergod** [comment]
Let's follow up on what we wanna fix or improve.

**Pull request merged**
